### PR TITLE
Kiota powered, in browser, client generation

### PR DESCRIPTION
**Pull request by andreaTP**
Those changes are completely in the UI, you can check them out using this command:

```
docker run --rm -it -p 8080:8080 docker.io/andreatp/apicurio-registry-kiota:latest
```

The Kiota compilation to Wasm and the production of the relevant folder happens in this repository:
https://github.com/andreaTP/apicurio-client-gen-poc

we might want to make it more official if we want to fully support it.

### CODE REVIEW

File: ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+                className="generate-client pf-m-redhat-font"
+                actions={[
+                    <Button key="Generate" variant="primary" data-testid="modal-btn-edit" onClick={this.doGenerate} isDisabled={!this.isGenerateEnabled()}>Generate</Button>,
+                    this.downloadLink()
+                ]}
+            >
+                <Form>
+                    <Grid hasGutter md={6}>
+                        <GridItem span={12}>
+                            <Label>[EXPERIMENTAL] This is "in-browser" generation of the client code using&nbsp;<a href="https://github.com/microsoft/kiota">Kiota</a>, please refer to&nbsp;
```
Comment (baywet) [inline ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx:98]: ❤️

File: ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+                                <Dropdown
+                                    toggle={
+                                        <DropdownToggle id="form-type-toggle" data-testid="form-type-toggle" onToggle={this.onLanguageToggle} toggleIndicator={CaretDownIcon}>
+                                            { this.state.data.language ? this.state.data.language : "Java" }
+                                        </DropdownToggle>
+                                    }
+                                    onSelect={this.onLanguageSelect}
+                                    isOpen={this.state.languageIsExpanded}
+                                    dropdownItems={[
+                                        <DropdownItem id="Java" key="Java" data-testid="form-type-auto"><i>Java</i></DropdownItem>,
```
Comment (baywet) [inline ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx:155]: may I suggest to sort them alphabetically? Also Kiota has a maturity matrix for languages, you might want to feature that too.

File: ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+                            <Label>[EXPERIMENTAL] This is "in-browser" generation of the client code using&nbsp;<a href="https://github.com/microsoft/kiota">Kiota</a>, please refer to&nbsp;
+                            <a href="https://microsoft.github.io/kiota/get-started/">the official documentation</a>&nbsp;to get started.</Label>
+                        </GridItem>
+
+                        <GridItem span={12}>
+                            <FormGroup
+                                label="Client Class Name"
+                                fieldId="form-client-name"
+                            >
+                                <TextInput
```
Comment (baywet) [inline ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx:107]: I do think you should feature the include and exclude filters, or have some kind of endpoint selection mechanism. A big value proposition of Kiota is being able to generate specifically for the API surface the client cares about. Another aspect to that is that often some descriptions won't be the cleanest, and filtering out things that are failing is a quick way to get around and get things done.

File: ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+            },
+            languageIsExpanded: false,
+            isValid: true,
+            downloadLink: "",
+            isErrorVisible: false,
+            isDownloadLinkVisible: false,
+        };
+    }
+
+    private doGenerate = async (): Promise<void> => {
```
Comment (baywet) [inline ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx:190]: from a UX perspective, having some kind of indication something is happening in the background would be helpful I think.

**Comment by baywet** [comment]
here is a short screencap of the feature for those interested.

apicurio-integration

File: ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx (hunk 2 above)
Comment (andreaTP) [reply ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx]: done 👍

File: ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx (hunk 3 above)
Comment (andreaTP) [reply ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx:110]: done 👍

File: ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx (hunk 4 above)
Comment (andreaTP) [reply ui/src/app/pages/artifactVersion/components/modals/generateClientModal.tsx:236]: agree, added a spinner 👍

**Comment by andreaTP** [comment]
@baywet thanks for the review!
I hope I have addressed all your points, and I think it looks a bit nicer now 🙂

**Review by baywet — APPROVED** [review]

**Comment by andreaTP** [comment]
This would also need documentation @smccarthy-ie 🙏

**Comment by andreaTP** [comment]
Now available in the Operate First environment:
https://apicurio-registry-mt-ui-mt-apicurio-apicurio-registry.apps.smaug.na.operate-first.cloud/

**Comment by baywet** [comment]
@andreaTP now that Kiota is (almost) integrated with apicurio, I'm wondering whether the search command from kiota should be able to sign-in and search an apicurio instance to go full circle?

**Comment by andreaTP** [comment]
> I'm wondering whether the search command from kiota should be able to sign-in and search an apicurio instance to go full circle?

@baywet interesting idea! But I'm not 100% sure about how the integration can work here, a few questions:

- should the integration work only through [our managed service](https://access.redhat.com/documentation/en-us/red_hat_openshift_service_registry/1/guide/ea7b18cd-e335-4fa9-98a3-3a23855d6e6a) or also on other instances?
- how do you see the Kiota Search functionality working in an integrated environment?

A slightly diverging angle in the 2 projects is that Kiota concentrates on OpenAPI while here Registry does support a number of different formats, do you have any plan for supporting other/more formats or should the feature be limited only to OpenAPI?

**Comment by baywet** [comment]
started a new discussion [here](https://github.com/microsoft/kiota/discussions/2154) to avoid side-tracking this PR.

**Comment by EricWittmann** [comment]
We should decide how to get this merged.  It needs to be behind a feature flag at least.

**Comment by EricWittmann** [comment]
Just hiding it is fine I think.  Hidden by default (for now), with a config variable to enable it.  Then we can set the config variable to enabled in Operate First.

**Review by carlesarnal — APPROVED** [review]

**Comment by carlesarnal** [comment]
This is now green. Unless there are any objections, I'll merge this mid-next week.

**Comment by andreaTP** [comment]
YaY!

**Comment by EricWittmann** [comment]
LGTM ready to merge

**Pull request merged**